Rewrite RHEL 8 and CS8 image definitions using the new framework

achilleas-k opened the pull request:
Followup to #3120.

This is a rewrite of the `distro/rhel8/` package to use the new image definition framework that was so far used only for Fedora.
I've described and provided context for all changes in individual commit messages.

Some things differ from how I did them for RHEL 9 because I had the opportunity to reevaluate certain decisions.  For example, all image types are now defined in functions.  I think we could standardise this and make every `imageType` be generated by a function that takes a `distribution` argument for making conditional checks on distribution name and version number.  We could use even for image types that don't need these checks for consistency.

The most notable _new thing_ is the `resolvePackageNames()` function in `rhel8/imagetype.go` (see commit with subject `distro/rhel8: resolve package names for RHEL 8`).  This is an issue we saw coming but the proper solution requires a lot more work.  The issue is that, outside of the preset package sets for each distro and image, we also select specific packages outside of the distribution implementation, in the manifest generation code in `manifest/`.  If a package has a different name on different distributions, this wont work, and we ran into it with `python3-toml` being selected when containers are embedded in ostree commits, because the package is named `python3-pytoml` on RHEL 8.  I have a rough plan for how to solve this and it involves having stages report their own package requirements but as "features" rather than package names.  Each distribution would then be responsible for resolving each feature into a known package name.  Since we only ran into this with a single package so far, we can get away with a single function in RHEL 8, but I wrote it in a way that can be extended for now in case we run into more examples before any further redesigns.

One somewhat unrelated change in this PR is the removal of the HostDistro in the distro registry.  These have been deprecated for a while and only created extra noise because of the requirement that we have distro constructors with extra arguments that were ignored.

achilleas-k commented:
Installer and edge commit tests are failing.
The installer failure is reproducible locally, so I'll start looking into that first.
Still not sure what's wrong with the commit build.

ondrejbudai commented on internal/distro/rhel8/distro.go:
```diff
@@ -678,63 +591,97 @@ func (t *imageType) checkOptions(customizations *blueprint.Customizations, optio
 func New() distro.Distro {
-	return newDistro("rhel-8")
+	// default minor: create default minor version (current GA) and rename it
+	d := newDistro("rhel", 6)
```
Current GA is 8.7, I wonder if that's important though. I must admit that the code smells a bit. :sweat_smile:

achilleas-k replied:
I fixed the 9 -> 9.1 in a followup commit but not this one.  I'll fix it up.

I'm not too happy about the "initialise and rename" either.  I'm trying to make it easy to have an unversioned constructor point to the proper release without that decision being buried deep in the distro code.  Maybe it belongs in the distro registry instead?

achilleas-k replied:
Fixed.  But I added it in a new commit instead of squashing into this one to make the change stand out on its own.
I also changed the repo definitions for tests so that rhel-8 == rhel-87, both in `test/data/manifests/` and `test-case-generators/`.

ondrejbudai commented on internal/distro/rhel9/distro.go:
```diff
@@ -156,16 +125,19 @@ func (d *distribution) getDefaultImageConfig() *distro.ImageConfig {
 func New() distro.Distro {
-	return NewRHEL90()
+	// default minor: create default minor version (current GA) and rename it
+	d := newDistro("rhel", 0)
```
Same here, current GA is 9.1. Sorry, we are pushing these two fast. :D

thozza replied:
This is fixed now.

thozza reviewed (dismissed):
Nice work 👏 

I added a few comments, but functionally changes look good to me.

FWIW, I have a few design-related comments:

- I didn't check it thoroughly, but functions in `internal/distro/*/images.go` look identical to me and I would expect these to be deduplicated and moved to a common location / package, instead of being copied to distros. Maybe that's the plan, but it looked like something that could have been part of this PR, thus I'm mentioning it. (I'm not saying it should be made part of this PR at this point)
- I'm not really in favor of making `imageTypes` functions shared among various distros due to (all are my opinions):
  - It is a step away from ever allowing images to be defined not as a Go code, which is something that I think we'll have to consider in the long term. Having to recompile a binary to define an image is IMHO very inconvenient and user-unfriendly for anyone wanting to experiment and define their own image (like Fedora Spins do).
  - I think this will become a mess full of conditionals mixing all distros and variants as we would be adding support for new distros (RHEL-10 and later) as things will inevitably differ.
  - IMHO defining image types as functions makes sense for distros that are almost the same and CentOS Stream / RHEL are good (and probably the only) example of where it makes sense, but only within one major version.
  - IMHO figuring our what an image definitions consist of should be reasonably easy without having to generate a manifest. We are making it actually harder and these cross-distro functions would make it even harder.

achilleas-k commented on internal/distro/rhel8/ami.go:
```diff
@@ -359,14 +359,6 @@ func defaultEc2SapImageConfigX86_64(rd distribution) *distro.ImageConfig {
 }
 
-// common ec2 image build package set
-func ec2BuildPackageSet(t *imageType) rpmmd.PackageSet {
```
Thanks for catching this.  I'll rebase.

achilleas-k replied:
Turns out this is correct because the ec2 build root was used in Azure as well.
From current `main`: https://github.com/osbuild/osbuild-composer/blob/560756a16419aad4278103aed0950ca8a0236062/internal/distro/rhel8/azure.go#L462

thozza replied:
Fair enough, thanks for checking 😉

achilleas-k commented on internal/distro/rhel8/distro.go:
```diff
@@ -281,11 +281,14 @@ func newDistro(name string, minor int) *distribution {
-		UEFIVendor: rd.vendor,
-		BasePlatform: platform.BasePlatform{
-			ImageFormat: platform.FORMAT_GCE,
+	// TODO: review requirement for platform with overridden packages for GCE
```
ah, interesting.  Should we change it in this PR?
I'm leaning towards no.  As with the RHEL 9 rewrite, I'm trying to keep changes to a minimum during the rewrite and make more purposeful PRs with image definition changes later.

thozza replied:
I'm all for not making this PR bigger, since Github is already having issues while reviewing it 😅

achilleas-k commented on internal/distro/rhel8/imagetype.go:
```diff
@@ -413,8 +413,39 @@ func (t *imageType) PackageSetsNew(bp blueprint.Blueprint, options distro.ImageO
+// workaround to the issue of having packages specified outside of distros (in
+// internal/manifest/os.go), which should be distro agnostic. In the future,
+// this should be handled more generally.
+func resolvePackageNames(packages []string) []string {
```
Good point.  I was thinking ahead to the feature-based one, but it's better to have the name describe what it's doing, not what it might be doing in the future.

achilleas-k replied:
Done!

achilleas-k commented on internal/distro/rhel8/imagetype.go:
```diff
@@ -413,8 +413,39 @@ func (t *imageType) PackageSetsNew(bp blueprint.Blueprint, options distro.ImageO
+
+// Runs resolvePackageNames() on each package set's Include and Exclude list
+// and replaces package names.
+func resolvePackageSetChains(chains map[string][]rpmmd.PackageSet) map[string][]rpmmd.PackageSet {
```
maybe `overridePackageNamesSets`?

thozza replied:
It's definitely better, but I'm still not sure. Also don't have any good ideas 😞  What about `overridePackageNamesInSets` ? 🙄

achilleas-k replied:
Done!

achilleas-k commented:
> Nice work clap
> 
> I added a few comments, but functionally changes look good to me.
> 
> FWIW, I have a few design-related comments:
> 
> * I didn't check it thoroughly, but functions in `internal/distro/*/images.go` look identical to me and I would expect these to be deduplicated and moved to a common location / package, instead of being copied to distros. Maybe that's the plan, but it looked like something that could have been part of this PR, thus I'm mentioning it. (I'm not saying it should be made part of this PR at this point)

I think that's where we're headed, yes.  I initially copied them from RHEL 9 expecting to need to make changes, but it none were needed.  I'd also do this in a follow-up and see if all distros (even Fedora) can share the same image functions.  I seem to recall I needed to make some changes from Fedora to RHEL 9, so maybe that will require a bit of work.


> * I'm not really in favor of making `imageTypes` functions shared among various distros due to (all are my opinions):
>   
>   * It is a step away from ever allowing images to be defined not as a Go code, which is something that I think we'll have to consider in the long term. Having to recompile a binary to define an image is IMHO very inconvenient and user-unfriendly for anyone wanting to experiment and define their own image (like Fedora Spins do).
>   * I think this will become a mess full of conditionals mixing all distros and variants as we would be adding support for new distros (RHEL-10 and later) as things will inevitably differ.
>   * IMHO defining image types as functions makes sense for distros that are almost the same and CentOS Stream / RHEL are good (and probably the only) example of where it makes sense, but only within one major version.
>   * IMHO figuring our what an image definitions consist of should be reasonably easy without having to generate a manifest. We are making it actually harder and these cross-distro functions would make it even harder.

I re-read my original comment on the PR and I think I was unclear about the image definition functions.  My thinking wasn't to share the, e.g., `qcow2ImageType()` function among distros, just to make it a soft rule (or even enforce it through the API) that image types be defined by an image type function that takes a distro as argument, even when not needed.  This could be enforced by changing the `addImageTypes()` function to be

```go
func (a *architecture) addImageTypes(platform platform.Platform, imageTypes ...imageTypeFunc)
```

where

```go
type imageTypeFunc func(d distribution) imageType
```


We would still have different private functions for `qcow2` in Fedora, RHEL 8, and RHEL 9, but since we're very often switching options based on OS version (and sometimes other things like arch or `isRHEL()`), it would be a good general rule that an image type generator function always has access to this information through the argument.

thozza commented on internal/distro/rhel9/distro.go:
```diff
@@ -71,36 +70,6 @@ var defaultDistroImageConfig = &distro.ImageConfig{
-	}
-}
-
-func getRHELDistro(major, minor int) distribution {
```
Yeah, this is definitely something for the future 😉

achilleas-k commented:
Updates in recent [force-push](https://github.com/osbuild/osbuild-composer/compare/0b25500f9e9f49ffd7f74c5de5dd3382fcf30208..7310d802fd7525cbbb2234504c36e9100f3edb74):
- Addressed the comments.
- Fixed the installer payload package set.  I accidentally dropped it completely before, which is why the installer tests were failing.
- Added a test that verifies that the pipeline names defined on an image type match the ones in each generated manifest.  This test revealed some names that weren't correct, so those were fixed.  The pipeline name mismatch was the reason for the edge-commit failures in CI.  The cloud API couldn't properly read the build metadata to return the payload packages because the name of the payload pipelines were wrong.

achilleas-k commented:
I need to rebase on main because https://github.com/osbuild/osbuild-composer/pull/3130 changed some functions used in distros and, while it doesn't cause conflicts, it creates errors on rebase.

achilleas-k commented:
New [force push](https://github.com/osbuild/osbuild-composer/compare/7310d802fd7525cbbb2234504c36e9100f3edb74..41049059d0e8056a975186b4c88415c628816221) is for rebase on main and fixes from that rebase.  Also, added a commit at the top of the PR that regenerates all RHEL 9 manifests because the sources were changed in a previous, merged PR, but the manifests weren't all updated.

thozza reviewed (approved):
Looks good 👍  Great additional fixes and test 😉

supakeen commented on internal/distro/rhel9/distro.go:
```diff
@@ -126,7 +126,7 @@ func (d *distribution) getDefaultImageConfig() *distro.ImageConfig {
 
 func New() distro.Distro {
 	// default minor: create default minor version (current GA) and rename it
-	d := newDistro("rhel", 0)
```
Same thing as before, another sidenote; can we document somewhere that these values 'need to be changed' every new RHEL minor?

achilleas-k replied:
Yes, I want to figure out a better way to do this too.  I think it should be defined in the distro or the distro registry, but not anywhere else (like, I wouldn't delegate it to an external configuration).  But I think it should be somewhere where it's clearly visible, easy to change, and we should have rules for when to update it.  Perhaps some release automation/test to avoid releasing composer with, for example, RHEL 8.9 support in RHEL 8.9 but where `rhel-8 == rhel-8.8`.

supakeen reviewed (approved):
Looks good, minor questions in two or three places.

achilleas-k commented on internal/distro/distro_test.go:
```diff
@@ -145,3 +148,96 @@ func TestImageType_PackageSetsChains(t *testing.T) {
+					// Add ostree options for image types that require them
+					options.OSTree = distro.OSTreeImageOptions{
+						ImageRef:      imageType.OSTreeRef(),
+						URL:           "https://example.com/repo",
```
I like using `example.com` because that's what it's there for.  Clearly broken can be messy.  Sometimes something that is clearly broken becomes valid down the line or we might add some validation that breaks the test.  The latter is a smaller concern, we could just fix the test with a new url when it breaks.

achilleas-k commented on internal/distro/rhel8/distro.go:
```diff
@@ -125,7 +125,7 @@ func (d *distribution) getDefaultImageConfig() *distro.ImageConfig {
 func New() distro.Distro {
 	// default minor: create default minor version (current GA) and rename it
-	d := newDistro("rhel", 6)
+	d := newDistro("rhel", 7)
```
We had that before and on the one hand I agree that it looks weird to just call with a single value for the minor.  On the other it's equally weird to have a private function with a `major` argument that always gets called with the same value.  I don't know which is less weird.  Trying out one of the two here I guess.

thozza replied:
I suggested to have a common cross-distro functions to create `distribution` structures  for CentOS / RHEL / Fedora and in such case, specifying the minor and major versions would be a must. I like your suggestion, but would pass major and minor version as two values to not have to split it inside...

achilleas-k commented:
Thanks for the reviews and comments.  I'll note all the ideas for followups.
